java/java-impl-inspections/src/com/intellij/codeInspection/CastCanBeReplacedWithVariableInspection.java

```diff
@@ -116,7 +113,52 @@ private static PsiVariable findReplacement(@NotNull PsiMethod method,
       }
     }
 
-    return null;
+    List<PsiAssignmentExpression> narrowTypeFound =
+      SyntaxTraverser.psiTraverser(method)
+        .filter(PsiAssignmentExpression.class)
+        .filter(assignment -> EquivalenceChecker.getCanonicalPsiEquivalence().expressionsAreEquivalent(assignment.getLExpression(), operand))
+        .filter(assignment -> {
+          PsiExpression typeElement = assignment.getRExpression();
+          return typeElement != null && InstanceOfUtils.typeCompatible(typeElement.getType(), castType, operand);
+        })
+        .toList();
+    for (PsiAssignmentExpression occurrence : narrowTypeFound) {
+      if (!isAtRightLocation(expression,  expressionTextRange, occurrence)) continue;
+
+      final PsiLocalVariable variable = getVariable(occurrence);
+
+      if (variable != null &&
+          resolveHelper.resolveReferencedVariable(variable.getName(), expression) == variable &&
+          !isChangedBetween(castedVar, methodBody, occurrence, expression) &&
+          !isChangedBetween(variable, methodBody, occurrence, expression)) {
+        return variable;
+      }
+    }
+
+    if (!(operand instanceof PsiReferenceExpression wideReferenceExpression)
+        || !(wideReferenceExpression.resolve() instanceof PsiLocalVariable wideVariable)) return null;
+
+    PsiExpression narrowExpression = wideVariable.getInitializer();
+    if (!(narrowExpression instanceof PsiReferenceExpression narrowReferenceExpression)
+        || !(narrowReferenceExpression.resolve() instanceof PsiVariable narrowVariable)
+        || !InstanceOfUtils.typeCompatible(narrowExpression.getType(), castType, operand)) return null;
+
+    if (!isAtRightLocation(expression, expressionTextRange, narrowExpression)) return null;
+
+    if (narrowVariable == null
```

Comment:
`null` check is redundant. The result of the `instanceof` operator is `false` if the value of the operand is `null`.

Reply:
Fixed